contracts/contracts/identity/KeyHolderLibrary.sol

```diff
@@ -143,14 +143,43 @@ library KeyHolderLibrary {
       return _keyHolderData.executionNonce-1;
   }
 
-  function removeKey(KeyHolderData storage _keyHolderData, bytes32 _key)
+  function removeKey(KeyHolderData storage _keyHolderData, bytes32 _key, uint256 _purpose)
       public
       returns (bool success)
   {
+      if (msg.sender != address(this)) {
+        require(keyHasPurpose(_keyHolderData, keccak256(abi.encodePacked(msg.sender)), 1), "Sender does not have management key"); // Sender has MANAGEMENT_KEY
+      }
+
       require(_keyHolderData.keys[_key].key == _key, "No such key");
-      emit KeyRemoved(_keyHolderData.keys[_key].key, _keyHolderData.keys[_key].purpose, _keyHolderData.keys[_key].keyType);
+      emit KeyRemoved(_key, _purpose, _keyHolderData.keys[_key].keyType);
+
+      // Remove purpose from key
+      uint256[] storage purposes = _keyHolderData.keys[_key].purposes;
+      for (uint i = 0; i < purposes.length; i++) {
```

Comment:
I originally had this as `purpose` instead of `purposes` in order to avoid these kind of loops and extra required storage on-chain (and associated gas costs). My idea was that you can represent multiple purposes with a single uint256 with each bit representing a different purpose, allowing us to represent 128 purposes for a single key. Since we only have 4 possible purposes right now, I figured 128 would be plenty. I also used this technique in the Marketplace contract to represent [multiple resolutions](https://github.com/OriginProtocol/origin-js/blob/master/contracts/contracts/marketplace/v01/Marketplace.sol#L270). Would be interested in your thoughts on that approach. I know it's different to the current standard so I should probably bring it up there too.

Reply:
Yeah I don't see why that can't be done to improve efficiency. I'm still thinking through the standard and trying to form my own opinions.

I'm not opposed to proposing a change to the standard. However, I do think we should be compliant with whatever the current standard is.

Reply:
I really liked @nick's proposal of using bitmasks for this. We should try and get it into the standard. Do we really think we need 128 different purposes? Seems like an uint8 ought to be enough.

Reply:
Ah, didn't know that, but it makes sense. Two uint128's seems more than adequate to me.